### PR TITLE
fix: improve error source detection

### PR DESCRIPTION
**Pull request by P0lip**
Fixes https://github.com/stoplightio/spectral/issues/658

**Checklist**

- [x] Tests added / updated
- [x] Docs added / updated

**Does this PR introduce a breaking change?**

- [ ] Yes
- [x] No

**Additional context**

We need to figure https://github.com/stoplightio/spectral/issues/608 out as soon as possible.
Basically, we need to make a call whether or not to return validation results that cannot be associated with any document, as we do now.
If we happen to do so, we'll keep reporting duplicate errors under certain circumstances.

### CODE REVIEW

File: src/resolved.ts [hunk 1]
```diff
@@ -45,13 +46,32 @@ export class Resolved {
...
+        } catch {
+          return false;
+        }
+
+        const { refMap } = nodeData;
+        let { $ref } = piece;
+
+        while ($ref in refMap) {
+          $ref = refMap[$ref];
+        }
```
Comment (marbemac) [inline src/resolved.ts]: This is an odd looking piece of code, could you add a comment describing what this is for?

Comment (marbemac) [reply src/resolved.ts]: Or just in general maybe this entire `hasRef` logic tree (line 55-74) could be pulled out to its own function, with a comment describing what it does (or some descriptive tests)?

File: src/utils/__tests__/refs.jest.test.ts [hunk 2]
```diff
@@ -0,0 +1,66 @@
+import { extractSourceFromRef, traverseObjUntilRef } from '..';
+
+describe('$ref utils', () => {
```
Comment (marbemac) [inline src/utils/__tests__/refs.jest.test.ts]: These might be a good fit for the JSON package (where we already have a bunch of pointer helpers etc)?

File: src/utils/hasRef.ts [hunk 3]
```diff
@@ -1,2 +1,5 @@
-export const hasRef = (obj: object): obj is object & { $ref: string } =>
-  '$ref' in obj && typeof (obj as Partial<{ $ref: unknown }>).$ref === 'string';
+import { isObject } from 'lodash';
+
+// needs to be in separate file due to rollup :/
+export const hasRef = (obj: unknown): obj is object & { $ref: string } =>
```
Comment (marbemac) [inline src/utils/hasRef.ts]: Another little json ref / pointer util could move to `json` package?

**Comment by nulltoken** [comment]
@marbemac @P0lip Did a quick test. No more URIError warns!  🎉

```bash
REDACTED@XXXXX MINGW64 /c/_work/spectral (fix/uri-error)
$ git log -1
commit a1cf6c206eb3cd11c4bdcb680bc54a382f13e61d (HEAD -> fix/uri-error, upstream/fix/uri-error)
Author: Jakub Rożek <jakub@stoplight.io>
Date:   Tue Dec 3 14:29:33 2019 +0100

    fix: normalize source

REDACTED@XXXXX MINGW64 /c/_work/spectral (fix/uri-error)
$ yarn cli lint --ruleset ./repro/ruleset.yaml ./repro/URIError.yaml
yarn run v1.19.2
$ node -r ts-node/register -r tsconfig-paths/register src/cli/index.ts lint --ruleset ./repro/ruleset.yaml ./repro/URIError.yaml
Missing baseUrl in compilerOptions. tsconfig-paths will be skipped
Function 'oasOpSecurityDefined' could not be loaded: Not Found
Function 'oasPathParam' could not be loaded: Not Found
Function 'oasOp2xxResponse' could not be loaded: Not Found
Function 'oasOpFormDataConsumeCheck' could not be loaded: Not Found
Function 'oasOpIdUnique' could not be loaded: Not Found
Function 'oasOpParams' could not be loaded: Not Found
Function 'refSiblings' could not be loaded: Not Found
Function 'oasTagDefined' could not be loaded: Not Found
OpenAPI 3.x detected

c:/_work/spectral/repro/URIError.yaml
 23:22  warning  schema-strings-maxLength  String typed properties MUST be further described using 'maxLength'. Error: paths./test.get.responses.200.content.application/json.schema.maxLength is not truthy

https://[REDACTED]]/static/schemas/common/v2/library.openapi.yaml
 65:15  warning  schema-strings-maxLength  String typed properties MUST be further described using 'maxLength'. Error: components.schemas.Error.properties.error.maxLength is not truthy
 65:15  warning  schema-strings-maxLength  String typed properties MUST be further described using 'maxLength'. Error: paths./test.get.responses.400.content.application/json.schema.properties.error.maxLength is not truthy
 67:27  warning  schema-strings-maxLength  String typed properties MUST be further described using 'maxLength'. Error: components.schemas.Error.properties.error_description.maxLength is not truthy
 67:27  warning  schema-strings-maxLength  String typed properties MUST be further described using 'maxLength'. Error: paths./test.get.responses.400.content.application/json.schema.properties.error_description.maxLength is not truthy
 69:21  warning  schema-strings-maxLength  String typed properties MUST be further described using 'maxLength'. Error: components.schemas.Error.properties.status_code.maxLength is not truthy
 69:21  warning  schema-strings-maxLength  String typed properties MUST be further described using 'maxLength'. Error: paths./test.get.responses.400.content.application/json.schema.properties.status_code.maxLength is not truthy

✖ 7 problems (0 errors, 7 warnings, 0 infos, 0 hints)

error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

However, I'm on a bit on the fence regarding the reported errors....

```
c:/_work/spectral/repro/URIError.yaml
 23:22  warning  schema-strings-maxLength  String typed properties MUST be further described
using 'maxLength'. Error: paths./test.get.responses.200.content.application/json.schema.maxLength
is not truthy

https://[REDACTED]]/static/schemas/common/v2/library.openapi.yaml
 65:15  warning  schema-strings-maxLength  String typed properties MUST be further 
described using 'maxLength'. Error: components.schemas.Error.properties.error.maxLength is 
not truthy
 65:15  warning  schema-strings-maxLength  String typed properties MUST be further
described using 'maxLength'. Error: paths./test.get.responses.400.content.application/json
.schema.properties.error.maxLength is not truthy
```

- From a "spec fixer" perspective, second [65:15] line reads a bit strange as this file doesn't contain any `paths` section. 
This path only makes sense in the context of the first file. But there's no hint to the file that indirectly suffers from this issue.
- If only the issue in `URIError.yaml` is fixed, my understanding is that Spectral will report the file as fine. Which is somehow wrong as it still references parts of a remote file containing issues.

Thoughts?

**Comment by P0lip** [comment]
> This path only makes sense in the context of the first file. But there's no hint to the file that indirectly suffers from this issue.

Yeaaah, I guess we'd need to provide a json path relevant to that file.

> If only the issue in URIError.yaml is fixed, my understanding is that Spectral will report the file as fine. Which is somehow wrong as it still references parts of a remote file containing issues.

It the referenced remote file contains issues, that error will still be reported.
Technically, URIError.yaml does not contain any errors, it's the remote that does.

**Comment by nulltoken** [comment]
@P0lip 
>Yeaaah, I guess we'd need to provide a json path relevant to that file.

:+1:

>Technically, URIError.yaml does not contain any errors, it's the remote reference that does.

I do agree.

>I suppose we could add a hint that URIError.yaml references a XYZ file that contains ABC errors.

That would also awesome. If that hint would be tied to some new built-in rule and that the user could change the severity of that rule, that would be even awesome-er ;-)

**Comment by P0lip** [comment]
cc @philsturgeon you may find the above convo quite interesting considering the Spectral VS Code plugin.

File: src/utils/__tests__/refs.jest.test.ts (hunk 2 above)
Comment (P0lip) [reply src/utils/__tests__/refs.jest.test.ts]: https://github.com/stoplightio/json/pull/38

**Comment by P0lip** [comment]
Alright, so I'm working on the path issue.
Pushed a commit https://github.com/stoplightio/spectral/commit/88eecf636aa942ea3c0aaf96c2811f2ce59f07d0.
Still needs to cleanup this up a little bit as well as update a few tests

**Comment by marbemac** [comment]
Not sure what the two files in question look like, but assuming the following:

`c:/_work/spectral/repro/URIError.yaml` has a ref at `paths./test.get.responses.400` that targets `https://[REDACTED]]/static/schemas/common/v2/library.openapi.yaml#/responses/bad-request`

I'd expect the resulting error report to look like:

```
https://[REDACTED]]/static/schemas/common/v2/library.openapi.yaml
 65:15  warning  schema-strings-maxLength  String typed properties MUST be further 
described using 'maxLength'. Error: components.schemas.Error.properties.error.maxLength is 
not truthy
 65:15  warning  schema-strings-maxLength  String typed properties MUST be further
described using 'maxLength'. Error: responses.bad-request.content.application/json
.schema.properties.error.maxLength is not truthy
```

Basically if we're reporting on another file, still reporting on inner paths relative to that file - in the example above it's the `#/responses/bad-request` root pointer that's being targeted on the target $ref.

Does that make sense / is that what you were thinking?

**Comment by nulltoken** [comment]
@marbemac 

> Not sure what the two files in question look like, but assuming the following:

You can the content of the files #658

**Pull request merged**